### ISSUE TITLE
🐛 BUG: Build/bundle assets not working

### ISSUE THREAD

**Issue by schlichtanders** [withastro/astro#2453]
### What version of `astro` are you using?

0.22.17

### What package manager are you using?

npm

### What operating system are you using?

Linux

### Describe the Bug

While this merged pullrequest https://github.com/withastro/astro/pull/1786 created some internal logic which copies assets, it unfortunately doesn't work.

The reason is mere file naming.
* Within the css files, `url(some/relative/path/to/myfile)` get translated to hashed `url(__VITE_ASSET_12314)`.
* The file on the other side, despite being copied to the correct location under dist/, is not renamed to `__VITE_ASSET_12314` but gets a combination of original name and hash, something like `myfile_XH14K`.

This makes it currently impossible to work with npm assets. It is as severe, that astro's internal logic will interfere (concretely overwrite it) other manual approaches like using `postcss-url` to accomplish the job of copying npm assets. That was quite surprising behaviour, seeing postcss-url copying files, which then get deleted again and renamed wrongly by some black astro magic.

TODO:
- [ ] to the best of my understanding, it is just the filenames of the copied assets which needs to be in sync with the pathname used in the transformed css. Everything else seems to be correctly in place.

------------------------

The MRE shows the use of fontsource. While `npm run dev` will show it correctly, running `npm run build && npm run preview` cannot find the font file as supposed to.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-gw1vrc-uazpyq?file=src/pages/index.astro

**Comment by nicobytes** [comment]
@schlichtanders I had the same problem and I didn't find the solution, finally I use astro-icon lib, but I have doubt what is the way for including CSS libraries in astro project.

**Comment by matthewp** [comment]
I'm going to take a quick look and see what's going on.

**Comment by matthewp** [comment]
I looked into this one and this is what's going on:

1. Astro has its own CSS plugin for bundling and doesn't use the builtin Vite one. This is because the builtin Vite one has private state that is needed for adding the correct <link> tags back into the rendered HTML.
1. Vite's HTML plugin has access to the private state, but this plugin is for SPA pattern with a single index.html, so we can't use it.
1. Vite also has an __asset__ plugin which we do use and that's the thing that marks up the `__VITE_ASSET__` in CSS.
1. This plugin also has private state that we would need that's shared with its own CSS plugin.

I was able to patch some changes locally to get the example app working. However it takes a change to Vite's API in order to expose the internal private state of the asset plugin. I'll reach out to the Vite team about exposing this so that we can use it. After that it will be a fairly simple fix.
